n_i2c.c

```diff
@@ -212,69 +212,108 @@ const char *i2cNoteTransaction(char *request, char **response, size_t timeoutMs)
 /**************************************************************************/
 bool i2cNoteReset()
 {
+    NOTE_C_LOG_DEBUG("resetting I2C interface...");
 
     // Reset the I2C subsystem and exit if failure
+    _DelayMs(CARD_REQUEST_I2C_SEGMENT_DELAY_MS);
     _LockI2C();
-    bool success = _I2CReset(_I2CAddress());
-    if (!success) {
+    if (!_I2CReset(_I2CAddress())) {
+        NOTE_C_LOG_ERROR(ERRSTR("unable to reset I2C interface.", c_err));
         _UnlockI2C();
         return false;
     }
-
-    // Synchronize by guaranteeing not only that I2C works, but that after we
-    // send \n that we drain the remainder of any pending partial reply from a
-    // previously-aborted session. If we get a failure on transmitting the \n,
-    // it means that the notecard isn't even present.
     delayIO();
-    const char *transmitErr = _I2CTransmit(_I2CAddress(), (uint8_t *)"\n", 1);
-    if (!cardTurboIO) {
-        _DelayMs(CARD_REQUEST_I2C_SEGMENT_DELAY_MS);
-    }
 
-    // This outer loop does retries on I2C error, and is simply here for robustness.
+    // The guaranteed behavior for robust resyncing is to send two newlines
+    // and  wait for two echoed blank lines in return.
     bool notecardReady = false;
-    int retries;
-    for (retries=0; transmitErr==NULL && !notecardReady && retries<3; retries++) {
+    for (size_t retries = 0; retries < CARD_RESET_SYNC_RETRIES ; ++retries) {
+
+        // Send a newline to the module to clean out request/response processing
+        // NOTE: This MUST always be `\n` and not `\r\n`, because there are some
+        //       versions of the Notecard firmware will not respond to `\r\n`
+        //       after communicating over I2C.
+        const char *transmitErr = _I2CTransmit(_I2CAddress(), (uint8_t *)"\n", 1);
+        // If we get a failure on transmitting the `\n`, it means that the
+        // Notecard isn't present.
+        if (transmitErr) {
+            NOTE_C_LOG_ERROR(transmitErr);
+            NOTE_C_LOG_ERROR(ERRSTR("failed to reset I2C", c_bad));

```

Comment:
Missing `_UnlockI2C()`.

Also, just to be clear, we are intentionally not retrying here, right? I'm curious why we don't retry on `_I2CTransmit` failing, but we do retry if `_I2CReceive` fails.

Reply:
I rewrote the function using TDD on the plane. I caught the `_UnlockI2C()` and, without thinking, I implemented the retry on transmit.

I just now ran the tests over the implementation you reviewed, and I completely agree. I have already refactored them to work just as you have suggested.